### PR TITLE
Add CUDF_TEST_PROGRAM_MAIN macro to tests lacking it

### PR DESCRIPTION
**Pull request by etseidl**
## Description
`JSON_PATH_TEST` and `ROW_CONVERSION_TEST` were not using the `CUDF_TEST_PROGRAM_MAIN`, and thus were not picking up the `GTEST_CUDF_RMM_MODE` env variable during nightly testing.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by etseidl** [comment]
CC @davidwendt

**Review by davidwendt — APPROVED** [review]

**Comment by vyasr** [comment]
It looks like with the change to using a pool you may need to up the GPU percentages used for both [Parquet](https://github.com/rapidsai/cudf/blob/branch-24.02/cpp/tests/CMakeLists.txt#L299) and [JSON](https://github.com/rapidsai/cudf/blob/branch-24.02/cpp/tests/CMakeLists.txt#L304). Maybe try 50 first and see if that works? I don't know the tests well enough to have any intuition for the data sizes.

**Comment by davidwendt** [comment]
> It looks like with the change to using a pool you may need to up the GPU percentages used for both [Parquet](https://github.com/rapidsai/cudf/blob/branch-24.02/cpp/tests/CMakeLists.txt#L299) and [JSON](https://github.com/rapidsai/cudf/blob/branch-24.02/cpp/tests/CMakeLists.txt#L304). Maybe try 50 first and see if that works? I don't know the tests well enough to have any intuition for the data sizes.

That is strange since neither of these two are part of those tests. 
Maybe the JSON_PATH_TEST and ROW_CONVERSION_TEST need their own GPU settings instead?

**Comment by etseidl** [comment]
> That is strange since neither of these two are part of those tests.
> Maybe the JSON_PATH_TEST and ROW_CONVERSION_TEST need their own GPU settings instead?

Did some local profiling, and it looks like `ROW_CONVERSION_TEST` blows up the memory pool right at the end (the last 1 or two tests it runs). I see that ROW_CONVERSION_TEST ends just after the two tests that have the memory error. If I run RCT with the cuda mr, I don't see the mem explosion. I'll have to look at the ManyStrings test...maybe it's fragmenting the pool memory?

**Comment by davidwendt** [comment]
> > That is strange since neither of these two are part of those tests.
> > Maybe the JSON_PATH_TEST and ROW_CONVERSION_TEST need their own GPU settings instead?
> 
> Did some local profiling, and it looks like `ROW_CONVERSION_TEST` blows up the memory pool right at the end (the last 1 or two tests it runs). I see that ROW_CONVERSION_TEST ends just after the two tests that have the memory error. If I run RCT with the cuda mr, I don't see the mem explosion. I'll have to look at the ManyStrings test...maybe it's fragmenting the pool memory?

The `ROW_CONVERSION_TEST` was recently moved from the JNI source tree to libcudf. My understanding is the move is temporary until the build team can figure out some CUDA 12 build issues within Spark. Given that, I would be ok with changes here to make it use less resources (and run faster as well).

**Comment by etseidl** [comment]
There are a bunch of things going on here. One is that the random number generator isn't seeded, so the order of tests, or selecting a subset, can lead to different results. For instance, on my machine, if I run just the `RowToColumnTests` suite, I get an OOM error in `ManyStrings` regardless of which memory resource I use. If I run the full set of tests, or just `ManyStrings`, no OOM.  If I add a `srand()` call, then no OOM.

But the real culprit is the `RowToColumnTests.Biggest` test which starts the explosive growth of the pooled MR.

> The `ROW_CONVERSION_TEST` was recently moved from the JNI source tree to libcudf. My understanding is the move is temporary until the build team can figure out some CUDA 12 build issues within Spark. Given that, I would be ok with changes here to make it use less resources (and run faster as well).

Well, if that's the case should we just revert to the default MR for that test? Or modify the tests to use less RAM?

**Comment by davidwendt** [comment]
> Well, if that's the case should we just revert to the default MR for that test? Or modify the tests to use less RAM?

I'd prefer to modify the tests to use less RAM? How feasible is that?

**Comment by vyasr** [comment]
> That is strange since neither of these two are part of those tests. Maybe the JSON_PATH_TEST and ROW_CONVERSION_TEST need their own GPU settings instead?

It could be (and sounds like) the case that the tests changed in this PR are now using up enough memory that other tests (the ones I mentioned) are failing. So yes, we could change the setting for those. OTOH if reducing the memory footprint of the `ROW_CONVERSION_TEST` is feasible that sounds even better.

**Comment by etseidl** [comment]
@ttnghia  can you take a look and see if it's ok to reduce the number of rows or columns in the big string tests?

**Comment by etseidl** [comment]
I reduced the row counts for two tests, and now the entire test stays within 6GB (when starting with a 3GB allocation).

**Comment by ttnghia** [comment]
> @ttnghia can you take a look and see if it's ok to reduce the number of rows or columns in the big string tests?

I'm not sure about this. @hyperbolic2346 what do you think?

File: cpp/tests/transform/row_conversion.cpp [hunk 1]
```diff
@@ -972,7 +975,7 @@ TEST_F(RowToColumnTests, ManyStrings)
     "this string is the longest string because it is duplicated more than you can imagine "
     "this string is the longest string because it is duplicated more than you can imagine "
     "this string is the longest string because it is duplicated more than you can imagine "
-    "this string is the longest string because it is duplicated more than you can imagine "
+    "this string is the longest string because it is duplicated more than you can imagine ",
     "a",
```
Comment (etseidl) [inline cpp/tests/transform/row_conversion.cpp:979]: I'm assuming "a" was not meant to be part of the preceding string.

Comment (davidwendt) [reply cpp/tests/transform/row_conversion.cpp:979]: Yes, it looks like it was meant to be it's own string.

File: cpp/tests/transform/row_conversion.cpp [hunk 2]
```diff
@@ -934,6 +935,8 @@ TEST_F(RowToColumnTests, BigStrings)
 
 TEST_F(RowToColumnTests, ManyStrings)
 {
+  // this test is very sensitive, so seed the random number generator
```
Comment (hyperbolic2346) [inline cpp/tests/transform/row_conversion.cpp:938]: Sensitive in what way?

Comment (etseidl) [reply cpp/tests/transform/row_conversion.cpp:938]: Prone to blowing up if too many instances of the largest string are used. Seeding the RNG prevents that.

Comment (etseidl) [reply cpp/tests/transform/row_conversion.cpp:920]: @hyperbolic2346 did you want a change to the comment here?

**Comment by hyperbolic2346** [comment]
> > @ttnghia can you take a look and see if it's ok to reduce the number of rows or columns in the big string tests?
> 
> 
> 
> I'm not sure about this. @hyperbolic2346 what do you think?

I think it is fine. The test sizes are somewhat arbitrarily chosen. It needs to be "large enough" to uncover bugs. Given this is temporary and 3 gigs is most likely large enough I am fine with the change.

**Review by hyperbolic2346 — APPROVED** [review]
Thanks!

**Pull request merged**